nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/list/ListedEntityTracker.java

```diff
@@ -0,0 +1,322 @@
+                    " For example, if set to '30 minutes', any entity having timestamp in recent 30 minutes will be the listing target when this processor runs." +
+                    " A listed entity is considered 'new/updated' and a FlowFile is emitted if one of following condition meets:" +
+                    " 1. does not exist in the already-listed entities," +
+                    " 2. has newer timestamp than the cached entity," +
+                    " 3. has different size than the cached entity." +
+                    " If a cached entity's timestamp becomes older than specified time window, that entity will be removed from the cached already-listed entities." +
+                    " %s", NOTE))
+            .addValidator(StandardValidators.TIME_PERIOD_VALIDATOR)
+            .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+            .defaultValue("3 hours")
+            .build();
+
+    private static final AllowableValue INITIAL_LISTING_TARGET_ALL = new AllowableValue("all", "All Available",
+            "Regardless of entities timestamp, all existing entities will be listed at the initial listing activity.");
+    private static final AllowableValue INITIAL_LISTING_TARGET_WINDOW = new AllowableValue("window", "Tracking Time Window",
+            "Ignore entities having timestamp older than the specified 'Tracking Time Window' at the initial listing activity.");
+
+    public static final PropertyDescriptor INITIAL_LISTING_TARGET = new PropertyDescriptor.Builder()
+            .name("et-initial-listing-target")
+            .displayName("Entity Tracking Initial Listing Target")
+            .description(format("Specify how initial listing should be handled." +
+                    " %s", NOTE))
+            .allowableValues(INITIAL_LISTING_TARGET_WINDOW, INITIAL_LISTING_TARGET_ALL)
+            .defaultValue(INITIAL_LISTING_TARGET_WINDOW.getValue())
+            .build();
+
+    public static final PropertyDescriptor NODE_IDENTIFIER = new PropertyDescriptor.Builder()
+            .name("et-node-identifier")
+            .displayName("Entity Tracking Node Identifier")
+            .description(format("The configured value will be appended to the cache key" +
+                    " so that listing state can be tracked per NiFi node rather than cluster wide" +
+                    " when tracking state is scoped to LOCAL. %s", NOTE))
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+            .defaultValue("${hostname()}")
+            .build();
+
+    private static Supplier<Long> CURRENT_TIMESTAMP = System::currentTimeMillis;
+
+    private Serializer<String> stringSerializer = (v, o) -> o.write(v.getBytes(StandardCharsets.UTF_8));
+    private Serializer<Map<String, ListedEntity>> listedEntitiesSerializer
+            = (v, o) -> objectMapper.writeValue(new GZIPOutputStream(o), v);
```

Comment:
When writing to a GZIPOutputStream, closing the stream is very important because it writes a trailer to the OutputStream. As far as I can tell, this stream is not being closed, so this is not really a proper GZIP stream. We should ensure that close the GZIPOutputStream. Also, by default, GZIPOutputStream uses a compression level of 5. We've done a huge amount of testing GZIPOutputStream against text and all results show that a compression level of 1 instead of 5 provides almost the same compression level (say, it gets us 98% of the way there) but compression level 1 is much less expensive. I'd recommend using a compression level of 1 as a result. Easiest way to do this is to use the `org.apache.nifi.stream.io.GZIPOutputStream`, which extends `java.util.zip.GZIPOutputStream` but allows you to provide a compression level.

Reply:
Thanks for catching this. I switched to use NiFi's GZIPOutputStream. As for closing stream, I don't think we can close the output stream passed to the serializer, because the stream may be used by the caller to write additional bytes. None of current implementation does that, but we can't be certain about if it's fine to close the stream here in this context.

GZIPOutputStream has `finish` method to finish writing Gzip data without closing the underlying stream. So I added calling finish from serializer.